chainindex/indexer.go

```diff
@@ -0,0 +1,450 @@
+	wg     sync.WaitGroup
+
+	db *sql.DB
+	cs ChainStore
+
+	idToRobustAddrFunc IdToRobustAddrFunc
+
+	insertEthTxHashStmt                   *sql.Stmt
+	getNonRevertedMsgInfoStmt             *sql.Stmt
+	getMsgCidFromEthHashStmt              *sql.Stmt
+	insertTipsetMessageStmt               *sql.Stmt
+	updateTipsetToRevertedStmt            *sql.Stmt
+	hasTipsetStmt                         *sql.Stmt
+	updateTipsetToNonRevertedStmt         *sql.Stmt
+	removeTipsetsBeforeHeightStmt         *sql.Stmt
+	removeEthHashesOlderThanStmt          *sql.Stmt
+	updateTipsetsToRevertedFromHeightStmt *sql.Stmt
+	updateEventsToRevertedFromHeightStmt  *sql.Stmt
+	isTipsetMessageNonEmptyStmt           *sql.Stmt
+	getMinNonRevertedHeightStmt           *sql.Stmt
+	hasNonRevertedTipsetStmt              *sql.Stmt
+	updateEventsToRevertedStmt            *sql.Stmt
+	updateEventsToNonRevertedStmt         *sql.Stmt
+	getMsgIdForMsgCidAndTipsetStmt        *sql.Stmt
+	insertEventStmt                       *sql.Stmt
+	insertEventEntryStmt                  *sql.Stmt
+
+	gcRetentionEpochs   int64
+	reconcileEmptyIndex bool
+	maxReconcileTipsets int
+
+	mu           sync.Mutex
+	updateSubs   map[uint64]*updateSub
+	subIdCounter uint64
+
+	closeLk sync.RWMutex
+	closed  bool
+}
+
+func NewSqliteIndexer(path string, cs ChainStore, gcRetentionEpochs int64, reconcileEmptyIndex bool,
+	maxReconcileTipsets int) (si *SqliteIndexer, err error) {
+	db, _, err := sqlite.Open(path)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to setup message index db: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer func() {
+		if err != nil {
+			_ = db.Close()
+			cancel()
+		}
+	}()
+
+	err = sqlite.InitDb(ctx, "chain index", db, ddls, []sqlite.MigrationFunc{})
+	if err != nil {
+		return nil, xerrors.Errorf("failed to init chain index db: %w", err)
+	}
+
+	si = &SqliteIndexer{
+		ctx:                 ctx,
+		cancel:              cancel,
+		db:                  db,
+		cs:                  cs,
+		updateSubs:          make(map[uint64]*updateSub),
+		subIdCounter:        0,
+		gcRetentionEpochs:   gcRetentionEpochs,
+		reconcileEmptyIndex: reconcileEmptyIndex,
+		maxReconcileTipsets: maxReconcileTipsets,
+	}
+	if err = si.prepareStatements(); err != nil {
+		return nil, xerrors.Errorf("failed to prepare statements: %w", err)
+	}
+
+	return si, nil
+}
+
+func (si *SqliteIndexer) Start() error {
+	si.wg.Add(1)
+	go si.gcLoop()
+	return nil
+}
+
+func (si *SqliteIndexer) SetIdToRobustAddrFunc(idToRobustAddrFunc IdToRobustAddrFunc) {
+	si.idToRobustAddrFunc = idToRobustAddrFunc
+}
+
+func (si *SqliteIndexer) Close() error {
+	si.closeLk.Lock()
+	defer si.closeLk.Unlock()
+	if si.closed {
+		return nil
+	}
+	si.closed = true
+
+	if si.db == nil {
+		return nil
+	}
+	si.cancel()
+	si.wg.Wait()
+
+	if err := si.db.Close(); err != nil {
+		return xerrors.Errorf("failed to close db: %w", err)
+	}
+	return nil
+}
+
+func (si *SqliteIndexer) prepareStatements() error {
```

Comment:
I'd strongly encourage you look at `preparedStatementMapping` in events/index.go for dealing with the 3 places that all of these have to appear in; reducing the duplication reduces the error-prone nature of setting these up and managing them, also reduces the amount of code needed to set these all up

Reply:
@rvagg I have made this change. This was a great refactor and definitely reduces boilerplate where things can go wrong 👍